ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,235 @@ type clientHandshakeState struct {
 	session         *ClientSessionState
 }
 
+type CacheKeyGenerator interface {
+	Key(net.Addr) string
+}
+
+type ClientFingerprintConfiguration struct {
+	// Version in the handshake header
+	HandshakeVersion uint16
+
+	// if len == 32, it will specify the client random
+	// Otherwise, the field will be random
+	ClientRandom []byte
+
+	// if RandomSessionID > 0, will overwrite SessionID w/ that many
+	// random bytes when a session resumption occurs
+	RandomSessionID int
+	SessionID       []byte
+
+	// These fields will appear exactly in order in the ClientHello
+	CipherSuites       []uint16
+	CompressionMethods []uint8
+	Extensions         []ClientExtension
+
+	// Optional, both must be non-nil, or neither.
+	// Custom Session cache implementations allowed
+	SessionCache ClientSessionCache
+	CacheKey     CacheKeyGenerator
+}
+
+type ClientExtension interface {
+	Marshal() []byte
+	CheckImplemented() error
+}
+
+func (c *ClientFingerprintConfiguration) CheckImplementedExtensions() error {
+	for _, ext := range c.Extensions {
+		if err := ext.CheckImplemented(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *ClientFingerprintConfiguration) ModifyConfig(config *Config) *Config {
+	config.NextProtos = []string{}
+	config.CipherSuites = c.CipherSuites
+	config.MaxVersion = c.HandshakeVersion
+	config.ClientRandom = c.ClientRandom
+	config.CurvePreferences = []CurveID{}
+	config.HeartbeatEnabled = false
+	config.ExtendedRandom = false
+	config.ForceSessionTicketExt = false
+	config.ExtendedMasterSecret = false
+	config.SignedCertificateTimestampExt = false
+	for i, ext := range c.Extensions {
+		switch casted := ext.(type) {
+		case SniExtension:
+			if casted.Autopopulate {
+				c.Extensions[i] = SniExtension{[]string{config.ServerName}, true}
```

Comment:
You should use named fields here, instead of relying on ordering. This also prevents compilation from breaking if you add a field.

```
c.Extensions[i] = SniExtension{
   Domains: []string{config.ServerName},
   Autopopulate: true,
}